ambry-store/src/main/java/com/github/ambry/store/BlobStore.java

```diff
@@ -660,15 +686,16 @@ public short undelete(MessageInfo info) throws StoreException {
       StoreKey id = info.getStoreKey();
       Offset indexEndOffsetBeforeCheck = index.getCurrentEndOffset();
       List<IndexValue> values = index.findAllIndexValuesForKey(id, null);
-      index.validateSanityForUndelete(id, values, IndexValue.LIFE_VERSION_FROM_FRONTEND);
+      index.validateSanityForUndelete(id, values, info.getLifeVersion());
       IndexValue latestValue = values.get(0);
       short lifeVersion = (short) (latestValue.getLifeVersion() + 1);
       MessageFormatInputStream stream =
           new UndeleteMessageFormatInputStream(id, info.getAccountId(), info.getContainerId(),
               info.getOperationTimeMs(), lifeVersion);
       // Update info to add stream size;
       info =
-          new MessageInfo(id, stream.getSize(), info.getAccountId(), info.getContainerId(), info.getOperationTimeMs());
+          new MessageInfo(id, stream.getSize(), info.getAccountId(), info.getContainerId(), info.getOperationTimeMs(),
+              info.getLifeVersion());
```

Comment:
Could you explain why we pass `info.getLifeVersion()` rather than resolved `lifeVersion` to message info?

Reply:
LifeVersion is used in two different places in undelete method. One is to construct an UndeleteMessageFormatInputStream, and we have to use a resolved lifeVersion, it can't be -1.
The other one is to verify the validity of a Undelete IndexEntry in PersistentIndex. when verifying undelete entry in persistent index (in method markAsUndelete), we need to pass lifeVersion as -1 if it's -1 in the message info so the persistentIndex knows it's from frontend.

Reply:
That's true.  I took a look at `writeTo(log)` method and found life version is not actually used when persisting to log. Then, it should be fine.